adding guide content (#1)
Laboratoria/guide-router: README.md

# Cómo codificar un router

## Índice

- [Qué es un Single Page Application (SPA)](#qué-es-un-single-page-application-spa)
- [Qué es un router](#qué-es-un-router)
- [Sobre window.location](#sobre-windowlocation)
- [Presentamos la History API](#presentamos-la-history-api)
- [Funcionalidades del router](#funcionalidades-del-router)
- [API de router básico](#api-de-router-básico)
- [Codificación de un SPA y un router](#codificación-de-un-spa-y-un-router)

Antes de comenzar a codificar un router, debemos cubrir un par de conceptos,
como la razón por la que existen los routers: la aplicación de página única
(Single Page Application - SPA).

> ## Qué es un Single Page Application (SPA)

Reviewer: ```suggestion
## ¿Qué es un Single Page Application (SPA)?
```

Author: omiti la punctuacion porque mi linter me dice que no podia tenerlo, pero creo mdlint pasa todo bien

Una aplicación de página única, en inglés Single Page Application o SPA,
es una aplicación web que utiliza un único archivo HTML
(probablemente `index.html`) y actualiza dinámicamente el contenido de ese
`index.html` a medida que el usuario interactúa con la aplicación.

En lugar de cargar páginas HTML separadas para cada navegación,
los SPA utilizan JavaScript para buscar y representar datos en la misma página
HTML, dando la ilusión de que estamos navegando para separar páginas HTML,
cuando en realidad es la misma página con contenido diferente renderizado
(dibujado).

> 

Reviewer: ```suggestion
Un Aplicación de Página Única, en inglés Single Page Application (SPA)
es una aplicación web que carga un único archivo HTML
(comúnmente llamado `index.html`) y actualiza dinámicamente el contenido
mientras el usuario interactúa con la aplicación.

En lugar de cargar páginas HTML separadas y completas cada vez que
se realiza una acción, como hacer clic en un enlace, solo se carga
el contenido necesario para actualizar la vista actual. Una SPA
da la ilusión de que estamos navegando por páginas HTML separadas,
pero en realidad es la misma página con contenido diferente renderizado
(dibujado) dinámicamente.
```

> ## Qué es un router

Reviewer: ```suggestion
## ¿Qué es un router?
```

En el contexto de los SPA, un router es un módulo de JavaScript que
gestiona la navegación dentro la aplicación.
Un router ayuda a crear una sensación de múltiples páginas o "vistas"
dentro de un SPA al manejar interacciones del usuario que provocan
cambios en la URL y la presentación del contenido apropiado.

Por ejemplo, si está en la página de inicio de un SPA `www.website.com` y
luego hace clic en un link del `<nav>` que te lleva a `www.website.com/about`,
el router se encarga de responder cambiando la URL, encontrar el contenido
apropiado para "about" y reemplazar el HTML en `index.html`
con el nuevo contenido. No existe un archivo `about.html` tal cual

> pero con el router y view parece que estamos navegando a una pagina nueva.

Reviewer: ```suggestion
En el contexto de los SPA, un router es un módulo de JavaScript que
gestiona la navegación dentro la aplicación sin necesidad de recargar
la página completa. Su función principal es mapear las URL a las
diferentes vistas o componentes de la aplicación y actualizar la interfaz
de usuario según la URL actual.

Por ejemplo, si un usuario está en la página de inicio de un SPA `www.website.com`,
luego hace clic en un link "About" que lo lleva a `www.website.com/about`,
el router detecta los cambios en la URL y encuentra y carga dinámicamente el
contenido correspondiente para "about". No existe un archivo `about.html` tal cual
pero con el router y view parece que estamos navegando a una pagina nueva.
Esto permite que la SPA simule la experiencia de navegación de una aplicación de
varias páginas, aunque en realidad solo haya cargado una única página HTML al
principio.
```

> 

Reviewer: ```suggestion

## Anatomía de una URL

Considera la siguiente URL de ejemplo:

``http://www.ejemplo.com:8080/pagina/ejemplo?clave=valor#seccion``

Las partes de esta URL son:

[protocol](https://developer.mozilla.org/en-US/docs/Web/API/URL/protocol):
indica el protocolo utilizado para acceder al recurso. En la URL de ejemplo es `http:`.

[host](https://developer.mozilla.org/en-US/docs/Web/API/URL/host):
indica la ubicación del servidor que aloja el recurso. Puede ser
una dirección IP o un nombre de dominio, en el ejemplo `www.ejemplo.com`.

[port](https://developer.mozilla.org/en-US/docs/Web/API/URL/port):
especifica el puerto del servidor al que se debe conectar. Si no
se especifica, se usa el puerto predeterminado para el protocolo
(80 para HTTP, 443 para HTTPS). En el ejemplo el puesto es  `8080`.

[pathname](https://developer.mozilla.org/en-US/docs/Web/API/URL/pathname):
especifica la ubicación del recurso en el servidor. En el
ejemplo, `/pagina/ejemplo`.

[search](https://developer.mozilla.org/en-US/docs/Web/API/URL/search):
contiene parámetros de búsqueda para la solicitud. Se inicia
con el símbolo de interrogación y tiene la forma
`clave=valor&clave2=valor2`.

[hash](https://developer.mozilla.org/en-US/docs/Web/API/URL/hash):
identifica una sección específica dentro del recurso. Se inicia
con el símbolo de numeral (#) seguido de un identificador. En la
URL de ejemplo, el hash es `seccion`

JavaScript proporciona el 
[`window.location`](https://developer.mozilla.org/es/docs/Web/API/Location)
a las diferentes partes de la URL actual de nuestra aplicación.
Consulta la documentacion para familiarizarte con estas propiedades.

```

> ### Entonces, qué son "routes" y "views"

Reviewer: ```suggestion
### Entonces, ¿qué son "routes" y "views"?
```

En su forma más simple, las rutas (routes) generalmente se definen como pares
`key-value`,

> donde `key` es la ruta o pathname de URL y el `value` es

Reviewer: ```suggestion
donde `key` es el pathname de URL y el `value` es
```

> una función asociada que representará la vista (view).

Reviewer: ```suggestion
una función asociada que genera la vista (view). Por ejemplo,
en el siguiente bloque de código, se define un objeto, donde
las llaves son las pathnames y los valores son funciones.
```

```js
const routes = {
  "/": Home,
  "/about": About
}
```

Las rutas se pueden definir de formas más complejas pero la idea base es

> la misma, un determinado ruta se relaciona con una determinada función de

Reviewer: ```suggestion
la misma, una determinada ruta se relaciona con una determinada función de
```

vista.

En este ejemplo, `Home` y `About` son funciones que, cuando se invocan,
construirán el html para la página y devolverá un elemento DOM que el

> router agregará al `index.html`.

Reviewer: ```suggestion
En este ejemplo, para la ruta "/" se debe usar la función `Home` y
para la ruta "/about" se debe usar la función `About`.
`Home` y `About`, cuando se invocan,
construirán la vista correspondiente y devolverán un elemento
[DOM](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement)
que el router agregará al `index.html`.
```

> 

Reviewer: ```suggestion

Por ejemplo, en el siguiente bloque de código, se define
una función `Home`, que crea, configura y retorna un <h1>.

```

```js
const Home = () => {
   const el = document.createElement("h1");
   el.innerHTML("I'm the Home Page");
  return el;
}
```

Entonces, si cargamos la página web y el nombre de la ruta es `/`,
el router lo sabrá por `routes["/"]` deberíamos invocar la función `Home`

> y veremos "I'm the Home Page".

Reviewer: ```suggestion
Si cargamos la página web y el pathname es `/` entonces
el router encontrará que debe invocar la función `Home`
e inyectará el <h1> que retorna y en la página web se
visualizará "I'm the Home Page".
```

O si cargamos la página web en `/about`, el router
sepa por `routes["/about"]` cómo llamar a la función `About`.

> Hay algunos pasos intermedios, pero esa es más o menos la idea.

Reviewer: Eliminar este caso. Creo que con Home queda claro.

```suggestion
```

## Sobre window.location

Ya que mencionamos `pathname`, revisemos brevemente `window.location`.
Con el objeto `window.location`, podemos acceder a partes de la URL
actual mediante programación.

Ubicación de ventana de ejemplo:

- `href`: `https://example.org:8080/foo/bar?q=baz#bang`
- `origin`: `https://example.org:8080`
- `port`: `8080`
- `pathname`: `/foo/bar`
- `search`: `?q=baz`
- `hash`: `#bang`

[Ver interactivamente partes de la URL](https://developer.mozilla.org/en-US/play)

Si no está familiarizado con `window.location` y sus propiedades
`origin`, `pathname`, `search`, ahora sería un buen momento para leer

> [los documentos de `location`.](https://developer.mozilla.org/es/docs/Web/API/Location)

Reviewer: Mover esto a la seccion que sugiero "Anatomia de  una URL"

```suggestion
```

Author: Me gusta la brevidad de esta seccion (habia algunas errores de traducion por google) - pero vamos porbando con la seccion que has escrito. Aprecio mucho el feedback.

# Cómo codificar un router

## Índice

> - [Qué es un Single Page Application (SPA)](#qué-es-un-single-page-application-spa)

Reviewer 2: ```suggestion
- [¿Qué es un Single Page Application (SPA?)](#qué-es-un-single-page-application-spa)
```

> - [Qué es un router](#qué-es-un-router)

Reviewer 2: ```suggestion
- [¿Qué es un router?](#qué-es-un-router)
```

- [Sobre window.location](#sobre-windowlocation)
- [Presentamos la History API](#presentamos-la-history-api)
- [Funcionalidades del router](#funcionalidades-del-router)
- [API de router básico](#api-de-router-básico)
- [Codificación de un SPA y un router](#codificación-de-un-spa-y-un-router)

Antes de comenzar a codificar un router, debemos cubrir un par de conceptos,
como la razón por la que existen los routers: la aplicación de página única
(Single Page Application - SPA).

## ¿Qué es un Single Page Application (SPA)?

Un Aplicación de Página Única, en inglés Single Page Application (SPA)
es una aplicación web que carga un único archivo HTML
(comúnmente llamado `index.html`) y actualiza dinámicamente el contenido
mientras el usuario interactúa con la aplicación.

En lugar de cargar páginas HTML separadas y completas cada vez que
se realiza una acción, como hacer clic en un enlace, solo se carga
el contenido necesario para actualizar la vista actual. Una SPA
da la ilusión de que estamos navegando por páginas HTML separadas,
pero en realidad es la misma página con contenido diferente renderizado
(dibujado) dinámicamente.

> ## ¿Qué es un router?

Reviewer 2: ```suggestion

## ¿Qué es un router?
```

En el contexto de los SPA, un router es un módulo de JavaScript que
gestiona la navegación dentro la aplicación sin necesidad de recargar
la página completa. Su función principal es mapear las URL a las
diferentes vistas o componentes de la aplicación y actualizar la interfaz
de usuario según la URL actual.

Por ejemplo, si un usuario está en la página de inicio de un SPA `www.website.com`,
luego hace clic en un link "About" que lo lleva a `www.website.com/about`,
el router detecta los cambios en la URL y encuentra y carga dinámicamente el

> contenido correspondiente para "about". No existe un archivo `about.html` tal cual

Reviewer 2: ```suggestion
contenido correspondiente para "about". No existe un archivo `about.html` como cual
```

pero con el router y view parece que estamos navegando a una pagina nueva.
Esto permite que la SPA simule la experiencia de navegación de una aplicación de
varias páginas, aunque en realidad solo haya cargado una única página HTML al

> principio.

Reviewer 3: ```suggestion
En el contexto de los SPA, un router es un módulo de JavaScript que
gestiona la navegación dentro de la aplicación sin necesidad de recargar
la página completa. Su función principal es mapear las URL a las
diferentes vistas o componentes de la aplicación y actualizar la interfaz
de usuario según la URL actual.

Por ejemplo, si un usuario que está en la página de inicio de un SPA `www.website.com`
hace clic en un link "About" que lo lleva a `www.website.com/about`,
el router va a detectar los cambios en la URL y encontrará y cargará dinámicamente el
contenido correspondiente para "about". No existe un archivo `about.html` tal cual,
pero con el router y view parece que estamos navegando a una pagina nueva.
Esto permite que la SPA simule la experiencia de navegación de una aplicación de
varias páginas, aunque en realidad solo haya cargado una única página HTML al
principio.
```

## Anatomía de una URL

Considera la siguiente URL de ejemplo:

``http://www.ejemplo.com:8080/pagina/ejemplo?clave=valor#seccion``

Las partes de esta URL son:

[protocol](https://developer.mozilla.org/en-US/docs/Web/API/URL/protocol):
indica el protocolo utilizado para acceder al recurso. En la URL de ejemplo es `http:`.

[host](https://developer.mozilla.org/en-US/docs/Web/API/URL/host):
indica la ubicación del servidor que aloja el recurso. Puede ser

> una dirección IP o un nombre de dominio, en el ejemplo `www.ejemplo.com`.

Reviewer 3: ```suggestion
una dirección IP o un nombre de dominio. En el ejemplo es `www.ejemplo.com`.
```

### Entonces, ¿qué son "routes" y "views"?

En su forma más simple, las rutas (routes) generalmente se definen como pares
`key-value`,
donde `key` es el pathname de URL y el `value` es
una función asociada que genera la vista (view). Por ejemplo,
en el siguiente bloque de código, se define un objeto, donde
las llaves son las pathnames y los valores son funciones.

```js
const routes = {
  "/": Home,
  "/about": About
}
```

Las rutas se pueden definir de formas más complejas pero la idea base es
la misma, una determinada ruta se relaciona con una determinada función de
vista.

En este ejemplo, para la ruta "/" se debe usar la función `Home` y
para la ruta "/about" se debe usar la función `About`.
`Home` y `About`, cuando se invocan,
construirán la vista correspondiente y devolverán un elemento
[DOM](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement)
que el router agregará al `index.html`.

Por ejemplo, en el siguiente bloque de código, se define
una función `Home`, que crea, configura y retorna un `<h1>`.

```js
const Home = () => {
   const el = document.createElement("h1");
   el.innerHTML("I'm the Home Page");
  return el;
}
```

> Si cargamos la página web y el pathname es `/` entonces
> el router encontrará que debe invocar la función `Home`
> e inyectará el `<h1>` que retorna y en la página web se
> visualizará "I'm the Home Page".

Reviewer 3: ```suggestion
Si cargamos la página web y el pathname es `/`,
el router encontrará que debe invocar la función `Home`
e inyectará el `<h1>` que retorna. Finalmente, en la página web se
visualizará "I'm the Home Page".
```

## Presentamos la History API

> A medida que navega por páginas web y carga una página tras otra, agrega
> contenido al historial de su navegador. Los botones de flecha hacia adelante
> y atrás en cada navegador le permiten avanzar y retroceder en su historial.
> 
> Con la History API podemos acceder y manipular el historial del navegador
> sin activar recargas de página completa. Un router en una SPA funciona
> junto con la History API para lograr esto. Carga de vista
> mientras se agrega al historial del navegador, preservando
> así la forma en que un usuario puede moverse atrás y adelante
> usando los botones en su navegador.
> 
> Si no fuera así, un SPA aún podría mostrar nuevas vistas
> en el mismo html, pero la URL probablemente no se actualizará
> y el historial nunca se agregará, por lo que presionar "back"
> al navegador saldría del sitio web por completo.
> 
> Los actores clave relacionados con la History API que
> utiliza un router son:
> 
> - [El método `pushState`](https://developer.mozilla.org/es/docs/Web/API/History/pushState)
>   nos permite agregar un nuevo estado a la cola del historial de la ventana
> - [El evento `popstate`](https://developer.mozilla.org/es/docs/Web/API/Window/popstate_event)
>   es un evento que la ventana se activa cuando cambia el historial.
>   Por ejemplo, cuando alguien presiona hacia atrás en el navegador.
> 
> Otros métodos de interés son `replaceState`, `go`, `forward`, `back`.
> Consulte [los documentos completos de History API](https://developer.mozilla.org/en-US/docs/Web/API/History_API)
> y [Trabajando con la History API](https://developer.mozilla.org/en-US/docs/Web/API/History_API/Working_with_the_History_API)
> 
> Otro evento de interés es `hashchange`, que no vamos a usar en nuestra router
> pero vale la pena saber. Consulte [los documentos de `hashchange`](https://developer.mozilla.org/en-US/docs/Web/API/Window/hashchange_event)
> para mas información.

Reviewer 3: ```suggestion
A medida que navegas por páginas web y cargas una página tras otra, se está agregando
contenido al historial de tu navegador. Los botones de flecha hacia adelante
y atrás en cada navegador permiten avanzar y retroceder en el historial.

Con la History API podemos acceder y manipular el historial del navegador
sin activar recargas de página completa. Un router en una SPA funciona
junto con la History API para lograr esto. Carga de vista
mientras se agrega al historial del navegador, preservando
así la forma en que un usuario puede moverse atrás y adelante
usando los botones en su navegador.

Si no fuera así, un SPA aún podría mostrar nuevas vistas
en el mismo html, pero la URL probablemente no se actualizará
y nunca se agregará al historial, por lo que presionar "back"
al navegador saldría del sitio web por completo.

Los actores clave relacionados con la History API que
utiliza un router son:

- [El método `pushState`](https://developer.mozilla.org/es/docs/Web/API/History/pushState)
  nos permite agregar un nuevo estado a la cola del historial de la ventana.
- [El evento `popstate`](https://developer.mozilla.org/es/docs/Web/API/Window/popstate_event)
  es un evento que la ventana se activa cuando cambia el historial.
  Por ejemplo, cuando alguien presiona hacia atrás en el navegador.

Otros métodos de interés son `replaceState`, `go`, `forward`, `back`.
Consulta [la documentación completa de History API](https://developer.mozilla.org/en-US/docs/Web/API/History_API)
y [Trabajando con la History API](https://developer.mozilla.org/en-US/docs/Web/API/History_API/Working_with_the_History_API).

Otro evento de interés es `hashchange`, que no vamos a usar en nuestro router
pero vale la pena conocerlo. Consulta [la documentación de `hashchange`](https://developer.mozilla.org/en-US/docs/Web/API/Window/hashchange_event)
para mas información.
```

## Funcionalidades del router

Un router básico debería:

> - Almacenar las rutas de la aplicación
> - Para un nombre de ruta determinado en la URL, muestre la vista
>   adecuada (y pasar los argumentos apropiados a la vista si hay
>   search parámetros)
> - Responder a la navegación dentro de la aplicación
>   (enlaces, clics en botones, etc.)
>   + agregando un nuevo estado al historial del navegador para actualizar la URL
>   + luego renderiza la vista apropiada según la ruta
>     (y pasar el argumentos apropiados si hay parámetros)
> - Responder a la navegación con los botones de avance y retroceso en el
>   navegador
>   + analizando la nueva URL para la ruta y los parámetros
>   + luego renderiza la vista apropiada según la ruta
>     (y pasar los argumentos apropiados si hay parámetros)
> - Cargando una página de error cuando la ruta no está definido en las rutas
> 
> No hay ningún error, mencionamos "renderizar la vista apropiada según la ruta"
> tres veces, entonces esto significa que escribiremos una función para esto
> para poder usarla varias veces.

Reviewer 3: ```suggestion
- Almacenar las rutas de la aplicación.
- Para un nombre de ruta determinado en la URL, renderizar la vista
  adecuada (y pasar los argumentos apropiados a la vista si hay
  search parámetros).
- Responder a la navegación dentro de la aplicación
  (enlaces, clics en botones, etc.)
  + agregando un nuevo estado al historial del navegador para actualizar la URL
  + luego renderizando la vista apropiada según la ruta
    (y pasar el argumentos apropiados si hay parámetros)
- Responder a la navegación con los botones de avance y retroceso en el
  navegador
  + analizando la nueva URL para la ruta y los parámetros
  + luego renderizando la vista apropiada según la ruta
    (y pasar los argumentos apropiados si hay parámetros)
- Cargando una página de error cuando la ruta no está definido en las rutas.

No hay ningún error, mencionamos "renderizar la vista apropiada según la ruta"
tres veces, entonces esto significa que escribiremos una función que haga esto
para poder usarla varias veces.
```

> Ahora profundizaremos explicando cómo lograr estas funcionalidades en código
> en `router.js`.

Reviewer 3: ```suggestion
Ahora profundizaremos explicando cómo lograr estas funcionalidades con código
en `router.js`.
```

### 1. Almacenar las rutas de la aplicación

El router necesita conocer las rutas de la aplicación para determinar qué vista
representar para una ruta.
Las rutas deben definirse en un módulo fuera del
router pero luego pasarse al router para almacenarlas.
El router debe tener su propia variable "privada" para almacenar las rutas
y una función que pueda llamarse desde otro módulo para asignar las rutas.

En código, eso significa que el router debe tener:

- `ROUTES` - objeto `{}`: Esta variable almacena información sobre
  las rutas en tu SPA.
- `setRoutes(routes)`: Esta función asigna el valor del parámetro
  `routes` al objeto `ROUTES`.

> ### 2. Para una ruta determinada en la URL, genere la vista adecuada
> 
> El router debe tener una función que, cuando se le da un `pathname`,
> muestre la vista. Entonces, ¿qué significa "muestra la vista"?
> 
> Si piensas en pseudocódigo, debemos borrar el html de la página actual,
> llame a la nueva función de vista correcta para renderizarla,
> y luego coloque el nuevo html en la página.

Reviewer 3: ```suggestion
### 2. Para una ruta determinada en la URL, generar la vista adecuada

El router debe tener una función que, cuando se le da un `pathname`,
muestre la vista. Entonces, ¿qué significa "generar la vista"?

Si piensas en pseudocódigo, debemos borrar el html de la página actual,
llamar a la nueva función de vista correcta para renderizarla,
y luego colocar el nuevo html en la página.
```

  programación a una nueva ruta dentro del SPA (por ejemplo, hacer clic en
  botones o enlaces dentro de la aplicación). Actualiza la URL usando
  `window.history.pushState` y luego llama a `renderView` con el `pathname`
  y los `props` dados para mostrar la vista correspondiente.

- `onURLChange(location)`: esta función está destinada a manejar cambios de URL
  con `popstate` y también cuando el app carga la primera vez.
  Cuando la URL cambia, analiza el `pathname` y la `search`
  desde la `location` (`window.location`) y luego llama a `renderView`.
  Debes usar `queryStringToObject` para obtener los parámetros de la
  cadena de `search` en formato de objeto para pasar a `renderView`.

## Codificación de un SPA y router

Armemos el router junto con algunas vistas sencillas para hacer un SPA!

### 1. Configura su estructura HTML

Crea un archivo HTML con la estructura básica de tu SPA.
Defina un elemento raíz (por ejemplo, un `<div id="root">`)
donde se representarán sus vistas.

```html
<!DOCTYPE html>
<html lang="en">
<head>
    <!-- Add your CSS and other head elements here -->
</head>
<body>
    <div id="root"></div> <!-- This is your root element -->

>     <script src="your-app-script.js"></script>

Reviewer 2: ```suggestion
    <script type="module" src="your-app-script.js"></script>
```

</body>
</html>
```

### 2. Crear las vistas "views"

Defina views. Un view es un componente función que representa una pagina entero.
Es una función que crea el contenido de la vista y debe devolver
un elemento DOM para el router usar.

Por ejemplo:

```js
// src/views/Home.js

> function Home(props) {

Reviewer 2: Tanto en ésta función como en el about, no se especifica que éstas funciones hay que exportarlas para poderlas importar desde otros archivos

> function Home(props) {

Author: ```suggestion
export function Home(props) {
```

  const viewEl = document.createElement('div');
  viewEl.textContent = 'Welcome to the home page!';
  return viewEl;
}

// src/views/About.js

> function About(props) {

Author: ```suggestion
export function About(props) {
```

  const viewEl = document.createElement('div');
  viewEl.textContent = 'This is the About page.';
  return viewEl;
}

// Definir funciones/componentes similares para otras rutas
```

### 3. Codifique el router

En su propio archivo `router.js`, implemente las partes del router siguiendo

> [la API descrita anteriormente](##API_básica_del_router).

Reviewer 2: ```suggestion
[la API descrita anteriormente](#api-de-router-básico).
```

export const setRootEl = (el) => {
  // assign rootEl
}

export const setRoutes = (routes) => {
  // optional Throw errors if routes isn't an object
  // optional Throw errors if routes doesn't define an /error route
  // assign ROUTES
}

const queryStringToObject = (queryString) => {
  // convert query string to URLSearchParams
  // convert URLSearchParams to an object
  // return the object
}

const renderView = (pathname, props={}) => {
  // clear the root element
  // find the correct view in ROUTES for the pathname
  // in case not found render the error view
  // render the correct view passing the value of props
  // add the view element to the DOM root element
} 

export const navigateTo = (pathname, props={}) => {
  // update window history with pushState
  // render the view with the pathname and props
}

export const onURLChange = (location) => {
  // parse the location for the pathname and search params
  // convert the search params to an object
  // render the view with the pathname and object
}
```

### 4. Configura el router

En su código JavaScript (por ejemplo, `index.js`), inicialice su router
definiendo sus rutas y configurando el elemento raíz:

```js

> import HomeView from './views/HomeView';

Reviewer 2: No se si esta línea confunda ya que antes se sugirió tener el nombre del archivo "Home.js"

> import HomeView from './views/HomeView';

Author: ```suggestion
import Home from './views/Home';
```

// ... import other views
import { setRootEl, setRoutes, onURLChange } from './router.js';

// Define your routes and their associated views
const routes = {

>   '/': HomeView,

Author: ```suggestion
  '/': Home,
```

  // ...
};

// Assign the routes
setRoutes(routes);

// Set the root element where views will be rendered
window.addEventListener("DOMContentLoaded", () => {
  setRootEl(/* root element */);
});
```

### 5. Manejar la carga de la primera página

Asegúrese de manejar la carga de la página inicial llamando a `onURLChange`
con `window.location`. 

En `index.js`

```js
// Handle initial URL load
window.addEventListener("DOMContentLoaded", () => {
  // set root element
  // invoke onURLChange 
});
```

Pruebe para ver si no importa con qué URL válida de su SPA comience,

> se carga la vista correcta.

Reviewer 2: ```suggestion
Pruebe para ver si, independientemente de qué URL válida se inicie su SPA,
se carga la vista correcta.
```

### 6. Implementar la navegación en la SPA

En sus vistas, puede utilizar enlaces de anclaje `<a>` o `<button>`
con `navigateTo` para navegar a diferentes rutas.
Recuerde que `navigateTo` debe tomar argumentos para `pathname` y `props`.

```js
// import navigateTo

> const Home = (props) => {

Author: ```suggestion
export const Home = (props) => {
```

  // ...
  linkEl.addEventListener('click', () => navigateTo("/about", { name: "Xochitl" }));
  // return el
}
```

### 7. Manejar hacia `forward` y `back`

Asegúrese de estar escuchando `popstate` en `index.js` y llame a la función
`onURLChange` cuando hay un `popstate`.

Cuando un usuario hace clic en un enlace o navega usando los botones
atrás/adelante del navegador, se llamará a esta función para actualizar
la vista mostrada.

```js
// Handle URL changes
window.addEventListener('popstate', ({objetivo}) => {
   onURLChange(/* location */);
});
```

Pruébelo usando los botones de avance y retroceso.

### 8. Probar la funcionalidad del `search` params

En una de sus vistas, experimente leyendo los parámetros de `search` de la
URL y utilizándolos en el vista. Las funciones de vista deben tener un
parámetro, llamémos `props`, que es un objeto con que podemos pasar información
a las vistas.

```js

> const Home = (props) => {

Author: ```suggestion
export const Home = (props) => {
```

# Cómo codificar un router

## Índice

- [¿Qué es un Single Page Application (SPA?)](#qué-es-un-single-page-application-spa)
- [¿Qué es un router?](#qué-es-un-router)
- [Sobre window.location](#sobre-windowlocation)
- [Presentamos la History API](#presentamos-la-history-api)
- [Funcionalidades del router](#funcionalidades-del-router)
- [API de router básico](#api-de-router-básico)
- [Codificación de un SPA y un router](#codificación-de-un-spa-y-un-router)

Antes de comenzar a codificar un router, debemos cubrir un par de conceptos,

> como la razón por la que existen los routers: la aplicación de página única
> (Single Page Application - SPA).

Reviewer 4: ```suggestion
como la razón por la que existen los routers: las Single Page Application
(aplicación de página única o SPA por sus siglas en inglés).
```

> ## ¿Qué es un Single Page Application (SPA)?

Reviewer 4: ```suggestion
## ¿Qué es una Single Page Application (SPA)?
```

Un Aplicación de Página Única, en inglés Single Page Application (SPA)

> es una aplicación web que carga un único archivo HTML

Reviewer 4: ```suggestion
es una aplicación web que carga todo su contenido a partir de un único archivo HTML
```

(comúnmente llamado `index.html`) y actualiza dinámicamente el contenido
mientras el usuario interactúa con la aplicación.

En lugar de cargar páginas HTML separadas y completas cada vez que
se realiza una acción, como hacer clic en un enlace, solo se carga
el contenido necesario para actualizar la vista actual. Una SPA
da la ilusión de que estamos navegando por páginas HTML separadas,
pero en realidad es la misma página con contenido diferente renderizado
(dibujado) dinámicamente.

## ¿Qué es un router?

> En el contexto de los SPA, un router es un módulo de JavaScript que

Reviewer 4: ```suggestion
En el contexto de los SPA, un router es un [`módulo`](https://es.javascript.info/modules-intro) de JavaScript que
```

gestiona la navegación dentro la aplicación sin necesidad de recargar

> la página completa. Su función principal es mapear las URL a las

Reviewer 4: ```suggestion
la página completa. Su función principal es asignar las URL a las
```

diferentes vistas o componentes de la aplicación y actualizar la interfaz
de usuario según la URL actual.

Por ejemplo, si un usuario está en la página de inicio de un SPA `www.website.com`,
luego hace clic en un link "About" que lo lleva a `www.website.com/about`,

> el router detecta los cambios en la URL y encuentra y carga dinámicamente el

Reviewer 4: ```suggestion
el router detecta los cambios en la URL, encuentra y carga dinámicamente el
```

> contenido correspondiente para "about". No existe un archivo `about.html` como cual

Reviewer 4: ```suggestion
contenido correspondiente para "about". No existe un archivo `about.html` como tal,
```

> pero con el router y view parece que estamos navegando a una pagina nueva.

Reviewer 4: ```suggestion
pero el router y view crean la sensación de que estamos navegando a una página nueva.
```

`key-value`,
donde `key` es el pathname de URL y el `value` es
una función asociada que genera la vista (view). Por ejemplo,
en el siguiente bloque de código, se define un objeto, donde
las llaves son las pathnames y los valores son funciones.

```js
const routes = {
  "/": Home,
  "/about": About
}
```

Las rutas se pueden definir de formas más complejas pero la idea base es
la misma, una determinada ruta se relaciona con una determinada función de
vista.

En este ejemplo, para la ruta "/" se debe usar la función `Home` y
para la ruta "/about" se debe usar la función `About`.
`Home` y `About`, cuando se invocan,
construirán la vista correspondiente y devolverán un elemento
[DOM](https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement)
que el router agregará al `index.html`.

Por ejemplo, en el siguiente bloque de código, se define
una función `Home`, que crea, configura y retorna un `<h1>`.

```js
const Home = () => {
   const el = document.createElement("h1");
   el.innerHTML("I'm the Home Page");
  return el;
}
```

Si cargamos la página web y el pathname es `/`,
el router encontrará que debe invocar la función `Home`

> e inyectará el `<h1>` que retorna. Finalmente, en la página web se
> visualizará "I'm the Home Page".

Reviewer 4: ```suggestion
tomará el  `<h1>` que esta función retorna y podremos agregarlo a nuestro archivo HTML. Finalmente, en la página web se
visualizará "I'm the Home Page".
```

> el router encontrará que debe invocar la función `Home`
> e inyectará el `<h1>` que retorna. Finalmente, en la página web se
> visualizará "I'm the Home Page".

Reviewer 4: ```suggestion
el router encontrará que debe invocar la función `Home` y tomar el  `<h1>` que esta función retorna para poder agregarlo a nuestro archivo HTML. Finalmente, en la página web se
visualizará "I'm the Home Page".
```

## Presentamos la History API

A medida que navegas por páginas web y cargas una página tras otra, se está agregando
contenido al historial de tu navegador. Los botones de flecha hacia adelante
y atrás en cada navegador permiten avanzar y retroceder en el historial.

Con la History API podemos acceder y manipular el historial del navegador
sin activar recargas de página completa. Un router en una SPA funciona

> junto con la History API para lograr esto. Carga de vista
> mientras se agrega al historial del navegador, preservando
> así la forma en que un usuario puede moverse atrás y adelante
> usando los botones en su navegador.

Reviewer 4: ```suggestion
junto con la History API para lograr esto. Carga la vista
mientras se agrega al historial del navegador, preservando de esta manera la forma en que un usuario puede moverse atrás y adelante
usando los botones en su navegador sin necesidad de cargar una nueva pagina
```

Esta función se puede importar a cualquier vista que tenga enlaces,
botones o necesite cargar una nueva vista y llamado en un detector de eventos.

```js
linkEl.addEventListener('click', () => navigateTo("/about", { name: "Xochitl" }))
```

#### Pasar argumentos a las vistas

Hablemos de esto - "y pasemos los argumentos apropiados a las vistas si hay" -
y por qué podría ser útil.

Hay ocasiones en las que queremos pasar datos a una vista, para que la misma
vista pueda renderizarse con información más específica dependiendo de qué
datos pasemos.

Por ejemplo, es posible que tengas una ruta para mostrar una vista de
información de la usuaria.

```js
  const routes = { "/profile": Profile }
```

Para cualquiera usuaria, queremos mostrar la misma página en general,
pero también con información específica, como el nombre de la usuaria en
particular.

En lugar de tener una ruta separada definida para cada posible
usuaria (tedioso y tal vez imposible) tenemos una ruta `/profile` que llama una
función de vista `Profile`, pero la función tiene parámetros para mostrar
algo diferente dependiendo del valor.

Ejemplo:

```js

> const Profile = ({ name }) => {

Author: ```suggestion
export const Profile = ({ name }) => {
```

de la `window`. Cuando se activa `popstate`, significa que la URL ha cambiado
debido a la navegación o interacción del usuario.

[Más información sobre `popstate`.](https://developer.mozilla.org/en-US/docs/Web/API/Window/popstate_event)

### 5. Cargando una página de error

Cuando el usuario carga una URL que no existe en la aplicación, la aplicación
debería mostrar una vista que le dice al usuario que hay un error de algún
tipo, o que no se encuentra.

Para apoyar esto comportamiento, las `routes` de una aplicación deben tener
una ruta con una vista definida para `/error` o `/not-found`.
En el momento en que el router va a renderizar la vista y busca
encontrar la vista para una ruta, si no existe ninguno, puede utilizar `/error`
como alternativa.

```js
const renderView = (pathname, props) => {
  // clear the current html
  find the correct view to render --or if there is none, use the error view--
  // call the view function (with the props) and get the new html element
  // put new html in the page
}
```

## API de router básico

Aquí hay una API completa de un router básico que tiene la funcionalidad
que acabamos de cubrir.

> - `routes` - objeto `{}`: Esta variable almacena información sobre las rutas en

Reviewer 4: Al ser elementos '_privados_' del archivo ROUTES y rootEl no forman parte de la API, en realidad solo se deberían mencionar sus setters como parte de la API

Author: Buen punto. Creo vale la pena mencionar los elementos privados para que ellas sepan que tiene que estar. Como planteamos?

Author: #2
